packages/flutter/lib/src/widgets/text_selection.dart

```diff
@@ -568,3 +569,143 @@ class _TextSelectionHandleOverlayState extends State<_TextSelectionHandleOverlay
     return null;
   }
 }
+
+/// A gesture detector to detect concurrent gesture events for a text field.
+///
+/// Unlike a plain [GestureDetector] where a single gesture wins the gesture
+/// arena, text fields supports chains of concurrent events.
+///
+/// See also:
+///
+///  * [TextField], a Material text field which uses this gesture detector.
+///  * [CupertinoTextField], a Cupertino text field which uses this gesture
+///    detector.
+class TextSelectionGestureDetector extends StatefulWidget {
+  /// Create a [TextSelectionGestureDetector].
+  ///
+  /// Multiple callbacks can be called from the same sequence of gesture.
+  /// The [child] parameter must not be null.
+  const TextSelectionGestureDetector({
+    this.onTapDown,
+    this.onSingleTapUp,
+    this.onSingleTapCancel,
+    this.onSingleLongTapDown,
+    this.onDoubleTapDown,
+    @required this.child,
+  }) : assert(child != null);
+
+  /// Called each distinct time a tap was held momentarily or on a short tap.
+  final GestureTapDownCallback onTapDown;
+
+  /// Called for each distinct short tap. Also called on the first tap of a
+  /// double tap but not on the second tap.
+  final GestureTapUpCallback onSingleTapUp;
+
+  /// Called for each tap down that does not become a short tap such as a long
+  /// tap or drag.
+  final GestureTapCancelCallback onSingleTapCancel;
+
+  /// Called for a single long tap that's sustained but not necessarily lifted.
+  /// Not called for a double-tap-hold.
+  final GestureLongPressCallback onSingleLongTapDown;
+
+  /// Called after a momentary hold or a short tap that is close in space and
+  /// time to a previous short tap.
+  final GestureTapDownCallback onDoubleTapDown;
+
+  /// Child below this widget.
+  final Widget child;
+
+  @override
+  State<StatefulWidget> createState() => _TextSelectionGestureDetectorState();
+}
+
+class _TextSelectionGestureDetectorState extends State<TextSelectionGestureDetector> {
+  // Is shortly after a previous single tap when not null.
+  Timer _doubleTapTimer;
+  Offset _lastTapOffset;
+  // True if second tap down of a double tap is detected. Used to discard
+  // subsequent tap up / tap hold of the same tap.
+  bool _isDoubleTap = false;
+
+  @override
+  void dispose() {
+    _doubleTapTimer?.cancel();
+    super.dispose();
+  }
+
+  // The down handler is force-run on success of a single tap and optimistically
+  // run before a long press success.
+  void _handleTapDown(TapDownDetails details) {
+    if (widget.onTapDown != null) {
+      widget.onTapDown(details);
+    }
+    // This isn't detected as a double tap gesture in the gesture recognizer
+    // because it's 2 single taps, each of which may do different things depending
+    // on whether it's a single tap, the first tap of a double tap, the second
+    // tap held down, a clean double tap etc.
+    if (_doubleTapTimer != null && _isWithinDoubleTapTolerance(details.globalPosition)) {
```

Comment:
If the double tap timer is running, then upon seeing the second tap down, I think it should always be reset.  Otherwise we could recognize an arbitrarily long sequence of taps as a double tap, so long as the intermediate ones were outside kDoubleTapSlop.

```
if (_doubleTapTimer != null) {
  if (widget.onDoubleTapDown != null && _isWithinDoubleTapTolerance(details.globalPosition)) {
     widget.onDoubleTapDown(details);
  }
  ...
}
```

Also: this logic seems like it would map a sequence of 4 quick taps to a 3 double taps? Doesn't seem right.

Reply:
We talked about this offline. Also added test